csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3620 @@
+
+    for (auto inp : input_groups) {
+      auto inp_promo_it = iel_promotion_map.find(inp);
+      if (inp_promo_it == iel_promotion_map.end()) {
+        promoted_inputs.push_back(inp->front());
+      } else {
+        promoted_inputs.push_back(inp_promo_it->second);
+        an_input_was_promoted = true;
+      }
+    }
+
+    if (!an_input_was_promoted) {
+      // No inputs need promotion so just continue
+      continue;
+    }
+
+    Expr* replay = nullptr;
+
+    auto promoted_input_groups = intersection_exact_loop_graph.toGroups(
+        VectorOfUniqueEntries<IterDomain*>{
+            promoted_inputs.begin(), promoted_inputs.end()});
+
+    // Before replaying, check if there's already an expression like this, if so
+    // use that for promotion. We would need the iel entries for non-promoted
+    // inputs to match exactly to reuse the expression.
+    ExprGroups non_promoted_input_uses;
+    for (auto iel_group : promoted_input_groups.intersect(input_groups)) {
+      non_promoted_input_uses.pushBack(
+          intersection_exact_loop_graph.uniqueUses(iel_group));
+    }
+
+    for (auto iel_use_group : non_promoted_input_uses) {
+      if (transformAtributesMatch(iel_expr->front(), iel_use_group->front())) {
+        auto use_inps =
+            ir_utils::filterByType<IterDomain>(iel_use_group->front()->inputs())
+                .vector();
+        bool inps_match = true;
+        for (auto inp_i : c10::irange(use_inps.size())) {
+          inps_match = inps_match &&
+              intersection_exact_loop_graph.disjointIdSets().strictAreMapped(
+                  use_inps[inp_i], promoted_inputs[inp_i]);
+        }
+        if (inps_match) {
+          replay = iel_use_group->front();
+          break;
+        }
+      }
+    }
+
+    if (replay == nullptr) {
+      replay = addReplayAs(promoted_inputs, iel_expr->front());
+      std::cout << "  ***REPLAY***:\n    " << iel_expr->front()
+                << "    As:" << replay->toString();
+    } else {
+      std::cout << "  Matched replay found: " << replay->toString();
+    }
+
+    auto out_groups = intersection_exact_loop_graph.outputGroups(iel_expr);
+
+    // Mark outputs as having a promoted iter domain
+    auto replay_out_ids =
+        ir_utils::filterByType<IterDomain>(replay->outputs()).vector();
+
+    TORCH_INTERNAL_ASSERT(replay_out_ids.size() == out_groups.size());
+
+    for (auto i : c10::irange(replay_out_ids.size())) {
+      iel_promotion_map[out_groups.vector()[i]] = replay_out_ids[i];
+    }
+  }
+
+  // Map from an exact iter domain group, to all the exact iter domain groups it
+  // covers
+  std::unordered_map<IdGroup, IdGroups> exact_covered_ids;
+
+  for (auto id_group :
+       idGraph(IdMappingMode::EXACT).disjointIdSets().disjointSets()) {
+    // Initialize inputs
+    if (idGraph(IdMappingMode::EXACT).uniqueDefinitions(id_group).empty()) {
+      exact_covered_ids[id_group] = {id_group};
+    }
+
+    // Initialize rfactor groups
```

Comment:
Why is this necessary?

Reply:
Because this analysis breaks down when we start trying to traverse across multiple rfactor domains. We really want to know which iter domains have the required resolved iter domain transformations in its history. To do so we check what contributions each iter domain has relative to one another.

If we have:

view [2, 8]->[4, 4]
and
view[8, 2]->[4, 4]

It's easy to compare the 4, 4 dimensions across the outputs, but comparing [2, 8] and [8, 2] doesn't make any sense because you're comparing overlapping domains, not disjoint ones.

Since view rfactor doesn't allow merging of iter domains and broadcast domains we don't have any resolutions between root and rfactor, so we simply stop when we hit rfactor transformations when looking in the history of what transformations are needed in the history of a leaf domain.

Reply:
But isn't it going to be overwritten by this?

```
for (auto output_group :
         idGraph(IdMappingMode::EXACT).outputGroups(exact_expr)) {
      exact_covered_ids[output_group] = covered;
    }
```

Reply:
Yeah you're right, no idea how this has been working. Maybe because I haven't been looking at a reshape case.